### PR TITLE
data source for a step function state machine

### PR DESCRIPTION
**Pull request by AndyMoore**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
New Data Source: aws_sfn_state_machine
```

Output from acceptance testing:

```
make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsSfnStateMachine'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsSfnStateMachine -timeout 120m
=== RUN   TestAccDataSourceAwsSfnStateMachine
=== PAUSE TestAccDataSourceAwsSfnStateMachine
=== CONT  TestAccDataSourceAwsSfnStateMachine
--- PASS: TestAccDataSourceAwsSfnStateMachine (32.05s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	32.094s

...
```

This is a data source for a SFN state machine. Only question is whether the additional attributes are needed beyond name/arn..

### CODE REVIEW

File: aws/data_source_aws_sfn_state_machine.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+func dataSourceAwsSfnStateMachine() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsSfnStateMachineRead,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"arn": {
```
Comment (bflad) [inline aws/data_source_aws_sfn_state_machine.go]: This attribute is missing documentation and is not implemented below (it should be though 👍 as its generally expected by Terraform AWS Provider users, even with the `id` attribute).

File: aws/data_source_aws_sfn_state_machine.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	sm, err := conn.DescribeStateMachine(&sfn.DescribeStateMachineInput{
+		StateMachineArn: aws.String(arns[0]),
+	})
+	if err != nil {
+		if awserr, ok := err.(awserr.Error); ok {
+			if awserr.Code() == "NotFoundException" || awserr.Code() == "StateMachineDoesNotExist" {
+				d.SetId("")
+				return nil
+			}
+		}
```
Comment (bflad) [inline aws/data_source_aws_sfn_state_machine.go]: This code should be removed from the data source, as it should error if the `DescribeStateMachine` call fails.

File: aws/data_source_aws_sfn_state_machine.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		return err
```
Comment (bflad) [inline aws/data_source_aws_sfn_state_machine.go]: We should return context about the error here for operators and code maintainers:

```suggestion
		return fmt.Errorf("error describing SFN State Machine (%s): %w, arns[0], err)
```

File: aws/data_source_aws_sfn_state_machine_test.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceAwsSfnStateMachineConfig(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsSfnStateMachineCheck(resourceName),
+					resource.TestCheckResourceAttrSet(resourceName, "role_arn"),
+					resource.TestCheckResourceAttrSet(resourceName, "definition"),
+					resource.TestCheckResourceAttrSet(resourceName, "creation_date"),
+					resource.TestMatchResourceAttr(resourceName, "creation_date", regexp.MustCompile("^20[0-9]{2}-")),
+					resource.TestCheckResourceAttr(resourceName, "status", "ACTIVE"),
```
Comment (bflad) [inline aws/data_source_aws_sfn_state_machine_test.go]: This data source testing can be more accurate to ensure the data source and resource values match by using `resource.TestCheckResourceAttrPair()`:

```go
					resource.TestCheckResourceAttrPair(resourceName, "arn", dataSourceName, "arn"),
					resource.TestCheckResourceAttrPair(resourceName, "creation_date", dataSourceName, "creation_date"),
					resource.TestCheckResourceAttrPair(resourceName, "definition", dataSourceName, "definition"),
					resource.TestCheckResourceAttrPair(resourceName, "name", dataSourceName, "name"),
					resource.TestCheckResourceAttrPair(resourceName, "role_arn", dataSourceName, "role_arn"),
					resource.TestCheckResourceAttrPair(resourceName, "status", dataSourceName, "status"),
```

File: aws/data_source_aws_sfn_state_machine_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceAwsSfnStateMachineCheck(name string) resource.TestCheckFunc {
```
Comment (bflad) [inline aws/data_source_aws_sfn_state_machine_test.go]: This function is extraneous with the usage of `resource.TestCheckResourceAttrPair()` noted above. 👍

File: aws/data_source_aws_sfn_state_machine.go (hunk 1 above)
Comment (AndyMoore) [reply aws/data_source_aws_sfn_state_machine.go]: added

File: aws/data_source_aws_sfn_state_machine_test.go (hunk 4 above)
Comment (AndyMoore) [reply aws/data_source_aws_sfn_state_machine_test.go]: added - cheers

File: aws/data_source_aws_sfn_state_machine_test.go (hunk 5 above)
Comment (AndyMoore) [reply aws/data_source_aws_sfn_state_machine_test.go]: 👍

**Comment by AndyMoore** [comment]
thanks for reviewing, think I got everything. still passing tests:

```make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsSfnStateMachine'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsSfnStateMachine -timeout 120m
=== RUN   TestAccDataSourceAwsSfnStateMachine
=== PAUSE TestAccDataSourceAwsSfnStateMachine
=== CONT  TestAccDataSourceAwsSfnStateMachine
--- PASS: TestAccDataSourceAwsSfnStateMachine (35.97s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	36.023s
```

File: aws/data_source_aws_sfn_state_machine_test.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+)
+
+func TestAccDataSourceAwsSfnStateMachine(t *testing.T) {
+	rName := acctest.RandString(5)
+	dataSourceName := "data.aws_sfn_state_machine.test"
+	resourceName := "data.aws_sfn_state_machine.test"
```
Comment (bflad) [inline aws/data_source_aws_sfn_state_machine_test.go]: Typo 😎 

```suggestion
	resourceName := "aws_sfn_state_machine.foo"
```

File: website/docs/d/sfn_state_machine.html.markdown [hunk 7]
```diff
@@ -0,0 +1,34 @@
+---
+subcategory: "SFN"
```
Comment (bflad) [inline website/docs/d/sfn_state_machine.html.markdown]: Looks like our documentation checker caught this:

```
Error checking Terraform Provider documentation: 1 error occurred:
	* website/docs/d/sfn_state_machine.html.markdown: error checking file frontmatter: YAML frontmatter subcategory (SFN) does not match allowed subcategories ([]string{"", "ACM PCA", "ACM", "API Gateway", "Access Analyzer", "AppMesh", "AppSync", "Application Autoscaling", "Athena", "Autoscaling", "Backup", "Batch", "Budgets", "Cloud9", "CloudFormation", "CloudFront", "CloudHSM v2", "CloudTrail", "CloudWatch", "CodeBuild", "CodeCommit", "CodeDeploy", "CodePipeline", "CodeStar Notifications", "Cognito", "Config", "Cost and Usage Report", "Data Lifecycle Manager (DLM)", "DataPipeline", "DataSync", "Database Migration Service (DMS)", "Device Farm", "Direct Connect", "Directory Service", "DocumentDB", "DynamoDB Accelerator (DAX)", "DynamoDB", "EC2", "ECR", "ECS", "EFS", "EKS", "ElastiCache", "Elastic Beanstalk", "Elastic Load Balancing (ELB Classic)", "Elastic Load Balancing v2 (ALB/NLB)", "Elastic Map Reduce (EMR)", "Elastic Transcoder", "ElasticSearch", "File System (FSx)", "Firewall Manager (FMS)", "Gamelift", "Glacier", "Global Accelerator", "Glue", "GuardDuty", "IAM", "Inspector", "IoT", "KMS", "Kinesis Firehose", "Kinesis", "Lambda", "License Manager", "Lightsail", "MQ", "Macie", "Managed Streaming for Kafka (MSK)", "MediaConvert", "MediaPackage", "MediaStore", "Neptune", "OpsWorks", "Organizations", "Pinpoint", "Pricing", "Quantum Ledger Database (QLDB)", "QuickSight", "RAM", "RDS", "Redshift", "Resource Groups", "Route53 Resolver", "Route53", "S3", "SES", "SNS", "SQS", "SSM", "SWF", "Sagemaker", "Secrets Manager", "Security Hub", "Service Catalog", "Service Discovery", "Service Quotas", "Shield", "SimpleDB", "Step Function (SFN)", "Storage Gateway", "Transfer", "VPC", "WAF Regional", "WAF", "WorkLink", "WorkMail", "WorkSpaces", "XRay"})
```

```suggestion
subcategory: "Step Function (SFN)"
```

**Comment by AndyMoore** [comment]
thanks - both fixed. tests still pass:

```
make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsSfnStateMachine'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsSfnStateMachine -timeout 120m
=== RUN   TestAccDataSourceAwsSfnStateMachine
=== PAUSE TestAccDataSourceAwsSfnStateMachine
=== CONT  TestAccDataSourceAwsSfnStateMachine
--- PASS: TestAccDataSourceAwsSfnStateMachine (35.29s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       35.348s
```

**Pull request merged**